File: AgoraChat/Code/Controllers/Threads/AgoraChatCreateThreadViewController.m
//
//  AgoraChatCreateThreadViewController.m
//  AgoraChat
//
//  Created by 朱继超 on 2022/4/5.
//  Copyright © 2022 easemob. All rights reserved.
//

#import "AgoraChatCreateThreadViewController.h"
#import "AgoraChatThreadListNavgation.h"
#import "AgoraChatUserDataModel.h"
#import "UserInfoStore.h"
#import "AgoraChatThreadViewController.h"

@interface AgoraChatCreateThreadViewController ()<EaseChatViewControllerDelegate>

@property (nonatomic) AgoraChatThreadListNavgation *navBar;

@property (nonatomic) EaseThreadCreateViewController *createViewController;

@property (nonatomic) AgoraChatGroup *group;

@end

@implementation AgoraChatCreateThreadViewController

- (instancetype)initWithType:(EMThreadHeaderType)type viewModel:(EaseChatViewModel *)viewModel message:(EaseMessageModel *)message {
    if ([super init]) {
        self.createViewController = [[EaseThreadCreateViewController alloc] initWithType:type viewModel:viewModel message:message];
        self.createViewController.delegate = self;
        [AgoraChatClient.sharedClient.groupManager getGroupSpecificationFromServerWithId:message.message.to completion:^(AgoraChatGroup *aGroup, AgoraChatError *aError) {
            if (!aError) {
                self.group = aGroup;
            }
        }];
    }
    return self;
}

- (void)didSendMessage:(AgoraChatMessage *)message thread:(nonnull AgoraChatThread *)thread error:(nonnull AgoraChatError *)error {
    EaseMessageModel *model = [[EaseMessageModel alloc]initWithAgoraChatMessage:message];
    if (!thread.threadId.length) {
        [self showHint:@"conversationId is empty!"];
        return;
    }
    model.thread = thread;
    AgoraChatThreadViewController *VC = [[AgoraChatThreadViewController alloc] initThreadChatViewControllerWithCoversationid:thread.threadId conversationType:AgoraChatConversationTypeGroupChat chatViewModel:self.createViewController.viewModel parentMessageId:thread.messageId model:model];
    VC.createPush = YES;
    VC.navTitle = thread.threadName;
    VC.detail = self.group.groupName;
    [self.navigationController pushViewController:VC animated:YES];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self _setupChatSubviews];
}

- (AgoraChatThreadListNavgation *)navBar {
    if (!_navBar) {
        _navBar = [[AgoraChatThreadListNavgation alloc]initWithFrame:CGRectMake(0, 0, EMScreenWidth, EMNavgationHeight)];
        _navBar.backgroundColor = [UIColor colorWithHexString:@"#FFFFFF"];
        __weak typeof(self) weakSelf = self;
        [_navBar setBackBlock:^{
            [weakSelf.navigationController popViewControllerAnimated:YES];
        }];
    }
    return _navBar;
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    self.navigationController.navigationBarHidden = YES;
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    self.navigationController.navigationBarHidden = NO;
}

- (void)_setupChatSubviews
{
    [self addChildViewController:self.createViewController];
    [self.view addSubview:self.navBar];
    self.navBar.title = @"New Thread";
    [self.view addSubview:self.createViewController.view];
    [self.createViewController.view mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.view).insets(UIEdgeInsetsMake(EMNavgationHeight, 0, 0, 0));
    }];
    self.view.backgroundColor = [UIColor colorWithRed:242/255.0 green:242/255.0 blue:242/255.0 alpha:1.0];
}

//userProfile
- (id<EaseUserProfile>)userProfile:(NSString *)huanxinID
{
    AgoraChatUserDataModel *model = nil;
    if ([huanxinID isEqualToString:@""] || huanxinID == nil) {
        return model;
    }
    AgoraChatUserInfo* userInfo = [[UserInfoStore sharedInstance] getUserInfoById:huanxinID];
    if(userInfo) {
        model = [[AgoraChatUserDataModel alloc]initWithUserInfo:userInfo];
    }else{
        [[UserInfoStore sharedInstance] fetchUserInfosFromServer:@[huanxinID]];
    }
    return model;
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Threads/AgoraChatCreateThreadViewController.swift
//
//  AgoraChatCreateThreadViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/25.
//

import UIKit

class AgoraChatCreateThreadViewController: UIViewController {

    private let navBar = AgoraChatThreadListNavgation()
    private let createViewController: EaseThreadCreateViewController
    private let message: EaseMessageModel
    private var group: AgoraChatGroup?
    
    init(type: EMThreadHeaderType, viewModel: EaseChatViewModel, message: EaseMessageModel) {
        self.message = message
        self.createViewController = EaseThreadCreateViewController(type: type, viewModel: viewModel, message: message)
        super.init(nibName: nil, bundle: nil)
        self.createViewController.delegate = self
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.view.backgroundColor = Color_F2F2F2
        
        self.navBar.frame = CGRect(x: 0, y: 0, width: UIScreen.main.bounds.width, height: 44 + self.view.safeAreaInsets.top)
        self.navBar.backgroundColor = .white
        self.navBar.backHandle = { [unowned self] in
            self.navigationController?.popViewController(animated: true)
        }
        self.navBar.title = "New Thread".localized
        self.view.addSubview(self.navBar)
        
        self.addChild(self.createViewController)
        self.view.addSubview(self.createViewController.view)
        
        self.navBar.snp.makeConstraints { make in
            make.left.right.top.equalTo(self.view)
            make.bottom.equalTo(self.createViewController.view.snp.top)
        }
        
        self.createViewController.view.snp.makeConstraints { make in
            make.left.right.bottom.equalTo(self.view)
            make.top.equalTo(self.view.safeAreaLayoutGuide).offset(44)
        }
        
        AgoraChatClient.shared().groupManager?.getGroupSpecificationFromServer(withId: self.message.message.to, completion: { group, error in
            self.group = group
        })
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.navigationController?.setNavigationBarHidden(true, animated: true)
    }
}

extension AgoraChatCreateThreadViewController: EaseChatViewControllerDelegate {
    func didSend(_ message: AgoraChatMessage, thread: AgoraChatThread, error: AgoraChatError?) {
        if thread.threadId.count <= 0 {
            self.showHint(String(format: "%@ can't be null".localized, "'conversationId'"))
            return
        }
        if let error = error {
            self.showHint(error.errorDescription)
            return
        }
        self.message.thread = thread
        let vc = AgoraChatThreadViewController(conversationId: thread.threadId, type: .groupChat, viewModel: self.createViewController.viewModel, parentMessageId: thread.messageId, model: self.message)
        vc.createPush = true
        vc.navTitle = thread.threadName
        vc.detail = self.group?.groupName
        self.navigationController?.pushViewController(vc, animated: true)
    }
    
    func userProfile(_ userID: String) -> EaseUserProfile? {
        if userID.count <= 0 {
            return nil
        }
        if let userInfo = UserInfoStore.shared.getUserInfo(userId: userID) {
            return AgoraChatUserDataModel(userInfo: userInfo)
        } else {
            UserInfoStore.shared.fetchUserInfosFromServer(userIds: [userID])
        }
        return nil
    }
}
